lib/srv/regular/sshserver.go

```diff
@@ -1096,40 +1110,221 @@ func (s *Server) getServerResource() (types.Resource, error) {
+
+		remoteFD, err := remote.File()
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+		defer remoteFD.Close()
+
+		_, _, err = proc.Conn.WriteWithFDs([]byte(addr), []*os.File{remoteFD})
+		if err != nil {
+			local.Close()
+			return nil, trace.Wrap(err)
+		}
+
+		return local, nil
+	}
+
+	// try to register with the parent context.
+	if other, ok := scx.Parent().TrySetDirectTCPIPForwardDialer(dialer); !ok {
+		// another forwarder was concurrently created. this isn't actually a problem, multiple forwarders
+		// being registered is harmless, but it does result in slightly higher resource utilization, so its
+		// preferable to use the existing forwarder and close ours in the background.
+		go proc.Close()
+		return other, nil
+	}
+
+	// successfully registered this dialer, add closer to context.
+	scx.Parent().AddCloser(proc)
+
+	return dialer, nil
+}
+
+// listenTCPIP creates a new listener in the forwarding process.
+func (s *Server) listenTCPIP(scx *srv.ServerContext, addr string) (net.Listener, error) {
+	proc, err := s.getTCPIPForwardProcess(scx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create a socket to receive new connections on.
+	localConn, remoteConn, err := uds.NewSocketpair(uds.SocketTypeDatagram)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer localConn.Close()
+	defer remoteConn.Close()
+	remoteFD, err := remoteConn.File()
+	if err != nil {
+		localConn.Close()
+		return nil, trace.Wrap(err)
+	}
+	defer remoteFD.Close()
+	_, _, err = proc.Conn.WriteWithFDs([]byte(addr), []*os.File{remoteFD})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The forwarding process may have chosen its own port, so we need to get the
+	// new listen address.
+	var fn int
+	fbuf := make([]*os.File, 1)
+	readDone := make(chan struct{})
+	// Read addr in another goroutine so we can cancel it if the forwarding process
+	// stops.
+	go func() {
+		_, fn, err = localConn.ReadWithFDs(nil, fbuf)
+		close(readDone)
+	}()
+
+	select {
+	case <-proc.Done:
+		localConn.Close()
+		return nil, trace.Errorf("forwarding process is closed")
+	case <-readDone:
+	}
+
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	defer listenerConn.Close()
+	if fn == 0 {
+		return nil, trace.BadParameter("forwarding process did not return a listener")
+	}
+	if err := validateListenerSocket(scx, localConn.UnixConn, fbuf[0]); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := net.FileListener(fbuf[0])
+	return listener, trace.Wrap(err)
+}
+
+func controlSyscallConn(conn syscall.RawConn, f func(fd uintptr) error) error {
+	var err error
+	err2 := conn.Control(func(fd uintptr) {
+		err = f(fd)
+	})
+	return trace.NewAggregate(err2, err)
+}
+
+// validateListenerSocket checks that the socket and listener file descriptor
+// sent from the forwarding process have the expected properties.
+func validateListenerSocket(scx *srv.ServerContext, controlConn *net.UnixConn, listenerFD *os.File) error {
```

Comment:
@jentfoo Let me know if there are any other properties I should be verifying in this function.

Reply:
@atburke I have not reviewed the code generally, but I took a close look at what we can validate here.

I have a possible expansion to the checks here in the commit: https://github.com/gravitational/teleport/commit/372ceb3a4f9246b0fffd4f630d28f48473f1570a

The commit describes the additional checks, as well as contains testing for these states.  Let me know if there are any questions or concerns in these checks, but generally I am just attempting to validate all the state we reasonably can for what we expect the socket to be.

If it looks good, feel free to cherry-pick the commit into your PR.

Reply:
From my understanding of the [man page](https://man7.org/linux/man-pages/man2/listen.2.html), I think the listener's socket is supposed to have type STREAM and not DGRAM? (everything else looks fine)

Reply:
I was unsure what type should be used as well.  Initially I expected `STREAM`, however I found your current implementation is using DGRAM: https://github.com/gravitational/teleport/pull/38828/files/07c9a5412ba5942f2abc5aea60fa968065c1e082#diff-5df50f8db69dd2760f95da564018976ba4467db43ea37c1f2f26cf7f8b22bd97R1172

So I updated the validation to require DGRAM to match the current implementation.  If we find STREAM is needed, we should just switch what we are expecting.

Reply:
That socket using DGRAM is the one that sends the listener's FD over, not the listener itself. I guess they're easy to mix up!

Reply:
Makes sense, thank you for cleaning up my commit!